### PR TITLE
feat(tianmu): add multi-thread aggregation of group by (#422)

### LINKED ISSUE #422: feature:  make aggregation parallel process to reduce execution time
### Documentation for mysql aggregation

 https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions-and-modifiers.html

## Functional Requirements:

1. Consistent with aggregation conditions handled by existing logic
2. The Aggregate function is used as the entry and ResultSender is used as the result output

### Table data limit:

1. The upper limit of a physical drive
2. The memory limit is exceeded. Procedure

### Memory usage limit:

1. Add parameters to control the maximum amount of memory that can be used for aggregation
2. Otherwise temporary disk files if the limit is exceeded

### Execution time limit:

1. Under the same conditions, it is in the same order of magnitude as InnoDB aggregate query at most
3. Aim to take less time than InnoDB for aggregated queries

 
##  Aggregate query  SQL

```sql

	select
		c_name,
		c_custkey,
		o_orderkey,
		o_orderdate,
		o_totalprice,
		sum(l_quantity)
	from
		customer,
		orders,
		lineitem
	where
		c_custkey = o_custkey
		and o_orderkey = l_orderkey
	group by
		c_name,
		c_custkey,
		o_orderkey,
		o_orderdate,
		o_totalprice
	order by
		o_totalprice desc,
		o_orderdate
	limit 10;

```

```sql
mysql> desc part;
+---------------+---------------+------+-----+---------+-------+
| Field         | Type          | Null | Key | Default | Extra |
+---------------+---------------+------+-----+---------+-------+
| p_partkey     | int(11)       | NO   | PRI | NULL    |       |
| p_name        | varchar(55)   | NO   |     | NULL    |       |
| p_mfgr        | char(25)      | NO   |     | NULL    |       |
| p_brand       | char(10)      | NO   |     | NULL    |       |
| p_type        | varchar(25)   | NO   |     | NULL    |       |
| p_size        | int(11)       | NO   |     | NULL    |       |
| p_container   | char(10)      | NO   |     | NULL    |       |
| p_retailprice | decimal(15,2) | NO   |     | NULL    |       |
| p_comment     | varchar(23)   | NO   |     | NULL    |       |
+---------------+---------------+------+-----+---------+-------+
9 rows in set (0.00 sec)


```

```sql

mysql> desc partsupp;
+---------------+---------------+------+-----+---------+-------+
| Field         | Type          | Null | Key | Default | Extra |
+---------------+---------------+------+-----+---------+-------+
| ps_partkey    | int(11)       | NO   | PRI | NULL    |       |
| ps_suppkey    | int(11)       | NO   | PRI | NULL    |       |
| ps_availqty   | int(11)       | NO   |     | NULL    |       |
| ps_supplycost | decimal(15,2) | NO   |     | NULL    |       |
| ps_comment    | varchar(199)  | NO   |     | NULL    |       |
+---------------+---------------+------+-----+---------+-------+
5 rows in set (0.00 sec)


```

```sql

    select
        p_brand,
        p_type,
        p_size,
        count(distinct ps_suppkey) as supplier_cnt
    from
        partsupp,
        part
    where
        p_partkey = ps_partkey
        and p_brand <> 'Brand#45'
        and p_type not like 'MEDIUM POLISHED%'
        and p_size in (49, 14, 23, 45, 19, 3, 36, 9)
        and ps_suppkey not in (
            select
                s_suppkey
            from
                supplier
            where
                s_comment like '%Customer%Complaints%'
        )
    group by
        p_brand,
        p_type,
        p_size
    order by
        supplier_cnt desc,
        p_brand,
        p_type,
        p_size
    limit 10;

```

### PR DESCRIPTION
**Pull request by adofsauron**
Documentation for mysql aggregation
    https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions-and-modifiers.html

    Functional Requirements:
    Consistent with aggregation conditions handled by existing logic
    The Aggregate function is used as the entry and ResultSender is used as the result output
    Performance Requirements:
    Table data limit:
    The upper limit of a physical drive
    The memory limit is exceeded. Procedure
    Memory usage limit:
    Add parameters to control the maximum amount of memory that can be used for aggregation
    Otherwise temporary disk files if the limit is exceeded
    Execution time limit:
    Under the same conditions, it is in the same order of magnitude as InnoDB aggregate query at most
    Aim to take less time than InnoDB for aggregated queries

## Summary about this PR

Issue Number: close #422

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [ ] Bug Fix
- [x] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/core/tianmu_table.cpp [hunk 1]
```diff
@@ -610,10 +610,9 @@ void TianmuTable::Field2VC(Field *f, loader::ValueCache &vc, size_t col) {
     case MYSQL_TYPE_SHORT:
     case MYSQL_TYPE_LONG:
     case MYSQL_TYPE_INT24:
-    case MYSQL_TYPE_LONGLONG: {
+    case MYSQL_TYPE_LONGLONG:
+    case MYSQL_TYPE_BIT: {
       int64_t value = f->val_int();
-      common::PushWarningIfOutOfRange(m_tx->Thd(), std::string(f->field_name), value, f->type(),
```
Comment (hustjieke) [inline storage/tianmu/core/tianmu_table.cpp]: Why rm this limit function? It will cause sth. wrong when we use unsigned data here.

**Review by hustjieke — CHANGES_REQUESTED** [review]

**Review by wisehead — APPROVED** [review]
@davidshiz has tested, no new bugs found, approving it.
https://stoneatom.yuque.com/staff-ft8n1u/hd9gr3/fgnonkvppw711h5k

File: mysql-test/suite/tianmu/r/out_of_range_issue1151.result [hunk 2]
```diff
@@ -22,10 +22,6 @@ insert into tiny values(128, 0);
 ERROR 22003: Out of range value for column 'a' at row 1
 insert into tiny values(1234, 0);
 ERROR 22003: Out of range value for column 'a' at row 1
-insert into tiny values(0, 128);
```
Comment (hustjieke) [inline mysql-test/suite/tianmu/r/out_of_range_issue1151.result:25]: Why deleted?

**Review by hustjieke — CHANGES_REQUESTED** [review]

File: mysql-test/suite/tianmu/r/out_of_range_issue1151.result (hunk 2 above)
Comment (adofsauron) [reply mysql-test/suite/tianmu/r/out_of_range_issue1151.result:25]: Temporary, as soon as this submission is merged

Comment (adofsauron) [reply mysql-test/suite/tianmu/r/out_of_range_issue1151.result:25]: Temporary, as soon as this submission is merged

**Review by RingsC — APPROVED** [review]

**Review by hustjieke — APPROVED** [review]
LGTM

**Pull request merged**